Add deepseek 3.2 exp

ArthurZucker opened the pull request:
# What does this PR do?

```python
from transformers import FineGrainedFP8Config, AutoModelForCausalLM, AutoTokenizer
import torch 


model_name = "deepseek-ai/DeepSeek-V3.2"
quantization_config = FineGrainedFP8Config(
    modules_to_not_convert=["model.layers.*.mlp.gate.*", "*.self_attn.indexer.weights_proj.*"],
    weight_block_size = (128, 128),

)
quantized_model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype="auto", device_map="auto", quantization_config=quantization_config)

tokenizer = AutoTokenizer.from_pretrained(model_name)
input_text = "What are we having for dinner?"
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda:0")

output = quantized_model.generate(**input_ids, max_new_tokens=10)
print(tokenizer.decode(output[0], skip_special_tokens=True))

```

nfywsh commented:
The submitted code is currently unusable and does not support the Deepseek-v3.2 official version. Is this PR still being updated?

jyliu24 commented:
> > The submitted code is currently unusable and does not support the Deepseek-v3.2 official version. Is this PR still being updated?
> 
> https://github.com/yunkchen/transformers/tree/v4.57.3_add_dpskv32

Isn't this implementation still O(L^2) since it just masks full attention to the indexer's topk?

freedom-cui commented:
@ArthurZucker @yunkchen 
hello， Has there been any progress?

RissyRan commented:
Hi @ArthurZucker, just checking in to see if there are any updates on this? There is a lot of interest for this change, so we’re excited to see it move forward! :) Thanks!

shuningjin commented:
Hi @ArthurZucker, Checking if there is any new updates. Look forward to the DeepSeek3.2 feature!

michaelroyzen commented:
+1 @ArthurZucker, getting this in would be extremely helpful

leideng commented:
Looking forward to the merge asap

ArthurZucker commented on docs/source/en/model_doc/deepseek_v32.md:
```diff
@@ -0,0 +1,63 @@
+
+The abstract from the paper is the following:
+
+<INSERT PAPER ABSTRACT HERE>
```
To do

ArthurZucker committed:
Add DeepSeek V3.2-Exp (DeepSeek Sparse Attention)

- DSA indexer key cache via DynamicIndexedLayer/StaticIndexedLayer (update_indexer)
- Add StaticIndexedLayer (static/compile counterpart); remove FP8DynamicIndexedLayer
- Register DynamicIndexedLayer + dynamic_sparse_attention causal mask
- Isolate glm_moe_dsa indexer diff; keep main's skip-topk feature
- Fill in deepseek_v32 docs

ArthurZucker committed:
Indexer returns the additive index mask directly (drop unused top-k indices)

The DSA indexer now builds and returns the [B, S, T] additive sparse mask itself,
instead of returning raw top-k indices that were only ever scattered into that mask.
GLM-MoE-DSA shares the index mask (not indices) across its skip-topk layers.

ArthurZucker commented on docs/source/en/model_doc/deepseek_v32.md:
```diff
@@ -0,0 +1,100 @@
+DSA reduces the quadratic cost of attention over long sequences by attending only to a selected subset of past tokens. It has two components:
+
+1. **Lightning indexer.** A lightweight, low-head-count scoring module computes an *index score* between each query and every preceding key. In the reference implementation it runs in FP8 with a Hadamard (`rotate_activation`) transform; because the transform is orthogonal (`Hq·Hk = q·k`) and FP8 is only a precision optimization, the transformers port computes the same scores directly in bf16/fp32 ...
+2. **Fine-grained token selection.** For each query the indexer keeps the top-`index_topk` (2048 by default) tokens, and main MLA attention is then computed only over those tokens via an additive mask. This turns the per-query attention cost from `O(L)` to `O(index_topk)` for long sequences.
```
```suggestion
2. **Fine-grained token selection.** For each query the indexer keeps the top-`index_topk` (2048 by default) tokens, and main MLA attention is then computed only over those tokens via an additive mask. This turns the per-query attention cost from `O(L)` to `O(index_topk)` for long sequences when using `flash_mla`, which is not supported yet 😉. 
```

ArthurZucker commented on docs/source/en/model_doc/deepseek_v32.md:
```diff
@@ -0,0 +1,100 @@
+In DeepSeek-V3.2 **every layer runs its own indexer** — there is no cross-layer top-k sharing.
+
+> [!TIP]
+> DeepSeek-V3.2 is the base DSA implementation in transformers. [`GlmMoeDsa`](./glm_moe_dsa) inherits from it and adds its own innovation on top: a per-layer `indexer_types` schedule (`"full"` vs `"shared"`) where `"shared"` layers skip running their own indexer and reuse the previous full layer's top-k selection, saving the per-layer indexer compute on long-context decoding (see [IndexCache](htt ...
```
```suggestion
```

ArthurZucker commented on docs/source/en/model_doc/deepseek_v32.md:
````diff
@@ -0,0 +1,100 @@
+print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+```
+
+This model was contributed by [Arthur](https://huggingface.co/ArthurZ).
````
```suggestion
```

ArthurZucker commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,594 @@
+        if past_key_values is not None:
+            k_cached = past_key_values.update_indexer(k, self.layer_idx)
+        else:
+            k_cached = k
```
```suggestion
        if past_key_values is not None:
            k = past_key_values.update_indexer(k, self.layer_idx)
```

ArthurZucker commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,594 @@
+
+        # === Scoring ===
+        # Reference: index_score = fp8_index(q_fp8, weights, k_cache, k_scale_cache). In bf16 mode the
+        # FP8 scale is 1, and `weights` already absorbs n_heads^(-0.5) and softmax_scale.
```
```suggestion
```

ArthurZucker commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,594 @@
+
+        self.scaling = self.qk_head_dim ** (-0.5)
+
+        # Every layer runs its own indexer (no cross-layer top-k sharing in DeepSeek V3.2).
```
```suggestion
```

ArthurZucker committed:
Make deepseek_v32 a lean DeepSeek-V3 + DSA model; glm_moe_dsa inherits it

- DeepseekV32Attention inherits DeepseekV3Attention (drops the duplicated MLA boilerplate);
  the indexer returns the additive index mask, folded into the attention mask, no flash padding.
- DeepseekV32Model/RMSNorm/RotaryEmbedding inherit DeepSeek-V3; rope_interleave=True.
- create_masks_for_generate returns a single mask when all layer_types match (full inheritance,
  no per-model dict-mask handling).
- Indexer scoring uses a plain weighted sum instead of einsum; RMSNorm sourced from DeepseekV3RMSNorm.
- Remove unused GatherParallel tensor-parallel style.

ArthurZucker committed:
Address review: matmul-only indexer, explicit attention forward, importable RoPE

- Indexer scoring uses matmul (no einsum) and reuses DeepSeek-V3's apply_rotary_pos_emb_interleave
  (drops the local single-tensor rope helper).
- DeepseekV32Attention writes its MLA forward explicitly (no super().forward), no dead flash padding.
- Simplify the indexer key-cache block (in-place k, no else branch); trim verbose comments.
- Docs: note the O(index_topk) speedup needs flash_mla (not supported yet).

ArthurZucker committed:
Revert DynamicCache __iter__ / ddp churn for indexer keys

Keep the cache_utils footprint to the essentials (DynamicIndexedLayer, StaticIndexedLayer,
update_indexer, dispatch); the indexer key cache is maintained via update_indexer during the
forward, so round-tripping it through the legacy/ddp tuple format is unnecessary.

ArthurZucker committed:
Docs: drop the GlmMoeDsa TIP and contributor line per review

molbap commented on src/transformers/models/deepseek_v32/configuration_deepseek_v32.py:
```diff
@@ -0,0 +1,144 @@
+    tie_word_embeddings: bool = False
+    rope_parameters: dict | None = None
+    # DeepSeek V3.2 applies RoPE in the interleaved (GPT-J pair) layout, like DeepSeek V3.
+    rope_interleave: bool = True
```
on that, maybe would be nice to address pending TODO in the apply rope interleave?

michaelroyzen commented:
Thank you so much for your continue work on this @ArthurZucker. Is it ready to merge?

ArthurZucker committed:
Gate DSA mask materialization to eager/SDPA; indexer returns top-k indices

The indexer now returns just the int32 top-k indices. Only the dense eager/SDPA paths
materialize the additive [-inf/0] sparse mask from them (so it stays visible to output
recorders); a flash-mla kernel would instead consume the indices directly (passed via
indices=), avoiding the redundant mask build. GLM threads the top-k indices across shared
layers (matching IndexCache), not a prebuilt mask.

molbap commented on tests/models/deepseek_v32/test_modeling_deepseek_v32.py:
```diff
@@ -0,0 +1,201 @@
+
+        tokenizer = AutoTokenizer.from_pretrained("deepseek-ai/DeepSeek-V2-Lite")
+        model = DeepseekV32ForCausalLM.from_pretrained(
+            "deepseek-ai/DeepSeek-V2-Lite",
```
to note, to change before merge (wrong model, same below)

molbap commented on src/transformers/conversion_mapping.py:
```diff
@@ -1329,6 +1329,14 @@ def _build_checkpoint_conversion_mapping():
+        WeightRenaming("indexer.wk", "indexer.k_proj"),
+        WeightRenaming("indexer.k_norm", "indexer.k_layernorm"),
+        WeightRenaming("indexer.wq_b", "indexer.q_b_proj"),
+        WeightRenaming("gate.e_score_correction_bias", "gate.bias"),
```
likely not to convert, seeing the module attributes?

ArthurZucker committed:
Simplify deepseek_v32 checkpoint conversion to the qwen2_moe base

The module now uses the checkpoint's own names (indexer wq_b/wk/k_norm/weights_proj, the MLA
projections, and gate.e_score_correction_bias), so the old indexer/gate renames are obsolete.
deepseek_v32 now uses the plain qwen2_moe expert-fusion mapping, like deepseek_v3 / glm_moe_dsa.

molbap commented on src/transformers/models/deepseek_v32/configuration_deepseek_v32.py:
```diff
@@ -0,0 +1,144 @@
+        if self.layer_types is None:
+            self.layer_types = ["dynamic_sparse_attention"] * self.num_hidden_layers
+        # Default to MoE from the second layer and on
+        if self.mlp_layer_types is None:
```
mlp layer types cannot be `None` here

molbap commented on src/transformers/conversion_mapping.py:
```diff
@@ -44,6 +44,7 @@
     "afmoe": "qwen2_moe",
     "deepseek_v2": "qwen2_moe",
     "deepseek_v3": "qwen2_moe",
+    "deepseek_v32": "qwen2_moe",
```
was commenting on that, perfect :D

molbap commented on src/transformers/cache_utils.py:
```diff
@@ -511,6 +587,73 @@ def reset(self):
+        cache_position = torch.arange(seq_len, device=self.indexer_device) + self.indexer_cumulative_length
+        # In-place to preserve the static data pointer (required for cudagraphs).
+        self.indexer_cumulative_length.add_(seq_len)
+        try:
+            self.indexer_keys.index_copy_(1, cache_position, indexer_key_states)
+        except NotImplementedError:
+            # Fallback for devices like MPS where index_copy_ might not be supported.
+            self.indexer_keys[:, cache_position] = indexer_key_states
```
would be better to catch & raise at init level no?

ArthurZucker replied:
No no because we don't know when it won't be supported / the op is just less optimized

molbap replied:
ah I see, indeed

ArthurZucker commented:
@michaelroyzen Just asked for a review, will be today I think!

ArthurZucker commented:
Follow up PR will be with the flash_mla interface

molbap commented on src/transformers/cache_utils.py:
```diff
@@ -274,6 +274,82 @@ def crop(self, max_length: int) -> None:
+        """
+        if not self.is_indexer_initialized:
+            self.lazy_initialization_indexer(indexer_key_states)
+        self.indexer_keys = torch.cat([self.indexer_keys, indexer_key_states], dim=1)
```
sounds good, no way to know ahead of time the final shape of indexer_key_states? I'm thinking how we can avoid cat but it's not a big deal either

molbap commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,400 @@
+        inv_freq = 1.0 / (
+            base ** (torch.arange(0, head_dim, 2, dtype=torch.int64).to(device=device, dtype=torch.float) / head_dim)
+        )
+        return inv_freq, attention_factor
```
a bit strange to have the early-return but same as v3 apart from that?

ArthurZucker replied:
yeah its just the `qk_rope_head_dim`

molbap commented on src/transformers/masking_utils.py:
```diff
@@ -1446,6 +1446,9 @@ def create_chunked_causal_mask(
     "heavily_compressed_attention": create_sliding_window_causal_mask,
+    # DSA (DeepSeek Sparse Attention) layers are causal at the mask level; the top-k token
+    # sparsification is applied separately inside the indexer, not via the layer mask.
+    "dynamic_sparse_attention": create_causal_mask,
```
"deepseek" I suppose too

ArthurZucker replied:
will see if I can fix but glm moe dsa used it I don't want to break it for them 👁️

molbap commented on src/transformers/models/glm_moe_dsa/modular_glm_moe_dsa.py:
much simpler

molbap commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,400 @@
+        cos, sin = position_embeddings
+
+        # === Queries ===
+        q = self.wq_b(q_resid)  # [B, S, H*D]
```
naming nits, do we have to keep these var names?

ArthurZucker replied:
we could change them for sure

molbap commented on src/transformers/models/deepseek_v32/modular_deepseek_v32.py:
```diff
@@ -0,0 +1,400 @@
+            index_scores = index_scores + attention_mask
+
+        topk = min(self.index_topk, index_scores.shape[-1])
+        # diff with classic attention, sample don't project :wink:
```
;)

molbap reviewed (commented):
Looking forward to this! left a small review, nothing jumps out so far!

ArthurZucker committed:
Fix CI: config-attributes check, indexer fp32 dtype + causality

- check_config_attributes: allowlist head_dim/layer_types/mlp_bias for the DSA configs; drop the
  now-unused rope_interleave field (the explicit forward always interleaves).
- Indexer: cast the weights_proj input to the layer's own dtype (it is kept in fp32 via
  _keep_in_fp32_modules but tests may force-cast), fixing fp16/static-cache crashes.
- Indexer causality: select top-k only at/before each query — use the passed mask when present
  (correct for padded/packed inputs), else reconstruct from position_ids (SDPA is_causal skip path).
  Fixes non-causal SDPA + static-cache.

ArthurZucker committed:
Make first_k_dense_replace configurable and fix tiny-config MoE test failures

- Add first_k_dense_replace config field (default 3) so the dense/MoE layer
  split is explicit; mlp_layer_types is derived from it in __post_init__.
- Tiny DeepseekV32 tester: first_k_dense_replace=1 gives a sparse MoE layer
  (so test_keep_in_fp32_modules_exist sees e_score_correction_bias) and
  n_group=1/topk_group=1 keeps group routing valid for 8 routed experts.
- Skip test_training_overfit (MoE on a tiny random model is unstable), matching glm_moe_dsa.
- Allowlist first_k_dense_replace in check_config_attributes.

ArthurZucker committed:
more efficient rotary pos emb

ArthurZucker committed:
manual cleanups

ArthurZucker committed:
up

ArthurZucker committed:
updates

ArthurZucker committed:
nits

ArthurZucker committed:
fix tests

ArthurZucker committed:
nits

ArthurZucker committed:
nit

ArthurZucker committed:
nits

ArthurZucker committed:
long context test

ArthurZucker committed:
remove the overwrite with a post_init fix?

ArthurZucker committed:
Merge branch 'add-deepseek-exp' of github.com:huggingface/transformers into add-deepseek-exp

ArthurZucker committed:
Tidy deepseek_v32 config comments

ArthurZucker committed:
Merge remote-tracking branch 'origin/main' into add-deepseek-exp

ArthurZucker committed:
add padded test and start styling

ArthurZucker committed:
nit

ArthurZucker committed:
Merge branch 'add-deepseek-exp' of github.com:huggingface/transformers into add-deepseek-exp

ArthurZucker commented:
run-slow: auto, deepseek_v3, deepseek_v32, glm4_moe_lite, glm_moe_dsa, longcat_flash, mistral4, youtu

ArthurZucker committed:
Apply ruff format to deepseek_v32 / glm_moe_dsa modular files

ArthurZucker committed:
Merge branch 'add-deepseek-exp' of github.com:huggingface/transformers into add-deepseek-exp

ArthurZucker committed:
Add publish/contribute dates to deepseek_v32 model card

vasqu committed:
Merge branch 'main' into add-deepseek-exp
